File: substrate/research_interrogation_subagent_chase_compose.py
"""Research interrogation → subagent chase compose (pure).

Operator vision: live in the research workstation; send subagents to chase
questions while interrogating/assessing/wrestling. Pure chase plan only.

live_dispatched always False.
pack_dispatched always False.
record_persisted always False.
prompts_injected always False.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Any, Literal

ChaseMode = Literal[
    "single_question",
    "swarm_fanout",
    "collective_merge_after",
]
VALID_MODES = frozenset(
    ("single_question", "swarm_fanout", "collective_merge_after")
)
VALID_FAMILIES = frozenset(
    ("arxiv", "substack", "openalex", "web", "custom")
)
_SECRETISH = re.compile(r"sk-|api[_-]?key|secret", re.I)


class ResearchInterrogationSubagentChaseComposeError(ValueError):
    """Fail-closed validation for interrogation subagent chase."""


@dataclass(frozen=True)
class PlannedChaseSlot:
    slot_id: str
    question_id: str
    body: str
    priority: int
    selected_model_id: str | None
    source_families: tuple[str, ...]
    live_dispatched: bool

    def to_dict(self) -> dict[str, Any]:
        return {
            "slot_id": self.slot_id,
            "question_id": self.question_id,
            "body": self.body,
            "priority": self.priority,
            "selected_model_id": self.selected_model_id,
            "source_families": list(self.source_families),
            "live_dispatched": False,
        }


@dataclass(frozen=True)
class ResearchInterrogationSubagentChaseCompose:
    session_id: str
    parent_asset_id: str
    chase_mode: str
    planned_slots: tuple[PlannedChaseSlot, ...]
    slot_count: int
    budget_ready: bool
    would_exceed: bool | None
    mark_for_twin_record: bool
    chase_ready: bool
    live_dispatched: bool
    pack_dispatched: bool
    record_persisted: bool
    prompts_injected: bool
    notes: tuple[str, ...]
    authority: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "session_id": self.session_id,
            "parent_asset_id": self.parent_asset_id,
            "chase_mode": self.chase_mode,
            "planned_slots": [s.to_dict() for s in self.planned_slots],
            "slot_count": self.slot_count,
            "budget_ready": self.budget_ready,
            "would_exceed": self.would_exceed,
            "mark_for_twin_record": self.mark_for_twin_record,
            "chase_ready": self.chase_ready,
            "live_dispatched": False,
            "pack_dispatched": False,
            "record_persisted": False,
            "prompts_injected": False,
            "notes": list(self.notes),
            "authority": "research_interrogation_subagent_chase_compose_advisory",
        }


def _require_nonempty(value: object, *, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise ResearchInterrogationSubagentChaseComposeError(
            f"{field} must be a non-empty string"
        )
    return value.strip()


def compose_research_interrogation_subagent_chase(
    *,
    session_id: object,
    parent_asset_id: object,
    questions: object,
    chase_mode: object,
    would_exceed: object,
    operator_ack: object,
    operator_override: object | None = None,
    selected_model_id: object | None = None,
    source_families: object | None = None,
    mark_for_twin_record: object | None = None,
) -> ResearchInterrogationSubagentChaseCompose:
    """Compose pure subagent chase plan. Never dispatches or persists."""
    if not isinstance(operator_ack, bool):
        raise ResearchInterrogationSubagentChaseComposeError(
            "operator_ack must be an explicit boolean"
        )
    if would_exceed is not None and not isinstance(would_exceed, bool):
        raise ResearchInterrogationSubagentChaseComposeError(
            "would_exceed must be boolean or null"
        )
    override = False if operator_override is None else operator_override
    if not isinstance(override, bool):
        raise ResearchInterrogationSubagentChaseComposeError(
            "operator_override must be boolean when set"
        )

    session = _require_nonempty(session_id, field="session_id")
    parent = _require_nonempty(parent_asset_id, field="parent_asset_id")
    if chase_mode not in VALID_MODES:
        raise ResearchInterrogationSubagentChaseComposeError(
            "chase_mode must be single_question|swarm_fanout|collective_merge_after"
        )
    if not isinstance(questions, list) or len(questions) == 0:
        raise ResearchInterrogationSubagentChaseComposeError(
            "questions must be a non-empty array"
        )

    mark = False if mark_for_twin_record is None else mark_for_twin_record
    if not isinstance(mark, bool):
        raise ResearchInterrogationSubagentChaseComposeError(
            "mark_for_twin_record must be boolean when set"
        )

    notes: list[str] = [
        "live_dispatched=false — chase plan is pure intent only",
        "pack_dispatched=false — collective merge after is intent only",
        "record_persisted=false — twin/session records not written",
        "prompts_injected=false — no live prompt mutation",
    ]

    model_id: str | None = None
    if selected_model_id is not None:
        model_id = _require_nonempty(
            selected_model_id, field="selected_model_id"
        )
        if len(model_id) > 128 or _SECRETISH.search(model_id):
            raise ResearchInterrogationSubagentChaseComposeError(
                "selected_model_id must be a model id, not secret material"
            )
        notes.append(f"selected_model_id={model_id} (operator authority)")
    else:
        notes.append(
            "selected_model_id=null — operator may choose before live chase"
        )

    families: list[str] = []
    if source_families is not None:
        if not isinstance(source_families, list):
            raise ResearchInterrogationSubagentChaseComposeError(
                "source_families must be an array when set"
            )
        seen_f: set[str] = set()
        for i, f in enumerate(source_families):
            if f not in VALID_FAMILIES:
                raise ResearchInterrogationSubagentChaseComposeError(
                    f"source_families[{i}] must be arxiv|substack|openalex|web|custom"
                )
            if f in seen_f:
                raise ResearchInterrogationSubagentChaseComposeError(
                    f"duplicate source_family: {f}"
                )
            seen_f.add(str(f))
            families.append(str(f))
    notes.append(f"source_family_count={len(families)}")

    normalized: list[dict[str, Any]] = []
    seen_q: set[str] = set()
    for i, q in enumerate(questions):
        if not isinstance(q, dict):
            raise ResearchInterrogationSubagentChaseComposeError(
                f"questions[{i}] must be an object"
            )
        qid = _require_nonempty(
            q.get("question_id"), field=f"questions[{i}].question_id"
        )
        body = _require_nonempty(q.get("body"), field=f"questions[{i}].body")
        if qid in seen_q:
            raise ResearchInterrogationSubagentChaseComposeError(
                f"duplicate question_id: {qid}"
            )
        seen_q.add(qid)
        priority = 0
        if "priority" in q and q["priority"] is not None:
            pr = q["priority"]
            if not isinstance(pr, int) or isinstance(pr, bool):
                raise ResearchInterrogationSubagentChaseComposeError(
                    f"questions[{i}].priority must be a finite integer"
                )
            priority = pr
        normalized.append(
            {"question_id": qid, "body": body, "priority": priority}
        )

    if chase_mode == "single_question" and len(normalized) != 1:
        raise ResearchInterrogationSubagentChaseComposeError(
            "single_question mode requires exactly 1 question"
        )
    if chase_mode in ("swarm_fanout", "collective_merge_after") and len(
        normalized
    ) < 2:
        raise ResearchInterrogationSubagentChaseComposeError(
            f"{chase_mode} requires ≥2 questions"
        )

    normalized.sort(
        key=lambda x: (-int(x["priority"]), str(x["question_id"]))
    )

    slots: list[PlannedChaseSlot] = []
    for idx, q in enumerate(normalized):
        slots.append(
            PlannedChaseSlot(
                slot_id=f"chase_{session}_{idx + 1}_{q['question_id']}",
                question_id=str(q["question_id"]),
                body=str(q["body"]),
                priority=int(q["priority"]),
                selected_model_id=model_id,
                source_families=tuple(families),
                live_dispatched=False,
            )
        )

    notes.append(
        f"planned_slots={len(slots)} · chase_mode={chase_mode}"
    )
    if chase_mode == "collective_merge_after":
        notes.append(
            "collective_merge_after — merge intent only; pack_dispatched=false"
        )
    if mark:
        notes.append(
            "mark_for_twin_record=true — candidates only; record_persisted=false"
        )

    budget_ready = False
    if would_exceed is None:
        if override:
            budget_ready = True
            notes.append(
                "budget_ready=true via operator_override (would_exceed unknown)"
            )
        else:
            notes.append(
                "budget_ready=false — would_exceed unknown and no operator_override"
            )
    elif would_exceed is True:
        if override:
            budget_ready = True
            notes.append(
                "budget_ready=true via operator_override despite would_exceed=true"
            )
        else:
            notes.append("budget_ready=false — would_exceed=true")
    else:
        budget_ready = True
        notes.append("budget_ready=true — would_exceed=false")

    chase_ready = operator_ack and budget_ready and len(slots) > 0
    if not operator_ack:
        notes.append("chase_ready=false — operator_ack required")
    elif not budget_ready:
        notes.append("chase_ready=false — budget gate closed")
    else:
        notes.append(
            "chase_ready=true — pure chase plan ready; still live_dispatched=false"
        )

    notes.extend(
        (
            "live_dispatched=false",
            "pack_dispatched=false",
            "record_persisted=false",
            "prompts_injected=false",
        )
    )

    return ResearchInterrogationSubagentChaseCompose(
        session_id=session,
        parent_asset_id=parent,
        chase_mode=str(chase_mode),
        planned_slots=tuple(slots),
        slot_count=len(slots),
        budget_ready=budget_ready,
        would_exceed=would_exceed if isinstance(would_exceed, bool) or would_exceed is None else None,
        mark_for_twin_record=mark,
        chase_ready=chase_ready,
        live_dispatched=False,
        pack_dispatched=False,
        record_persisted=False,
        prompts_injected=False,
        notes=tuple(notes),
        authority="research_interrogation_subagent_chase_compose_advisory",
    )


def format_research_interrogation_subagent_chase_summary(
    c: ResearchInterrogationSubagentChaseCompose,
) -> str:
    return (
        f"chase_ready={c.chase_ready} · mode={c.chase_mode} · "
        f"slots={c.slot_count} · budget_ready={c.budget_ready} · "
        f"live_dispatched=false · pack_dispatched=false · "
        f"record_persisted=false · prompts_injected=false"
    )


__all__ = [
    "PlannedChaseSlot",
    "ResearchInterrogationSubagentChaseCompose",
    "ResearchInterrogationSubagentChaseComposeError",
    "compose_research_interrogation_subagent_chase",
    "format_research_interrogation_subagent_chase_summary",
]

File: apps/reading/src/api/researchInterrogationSubagentChaseCompose.ts
/**
 * Research interrogation → subagent chase compose (pure).
 *
 * Operator vision: live in the research workstation; send subagents to chase
 * questions while interrogating/assessing/wrestling with information in front
 * of you. Pure plan of chase intents — never live-dispatches.
 *
 * live_dispatched always false.
 * pack_dispatched always false.
 * record_persisted always false.
 * prompts_injected always false.
 */

export type ChaseMode =
  | "single_question"
  | "swarm_fanout"
  | "collective_merge_after";

export type SourceFamilyHint =
  | "arxiv"
  | "substack"
  | "openalex"
  | "web"
  | "custom";

export interface ChaseQuestion {
  question_id: string;
  body: string;
  /** Optional higher = earlier in plan. */
  priority?: number;
}

export interface PlannedChaseSlot {
  slot_id: string;
  question_id: string;
  body: string;
  priority: number;
  selected_model_id: string | null;
  source_families: SourceFamilyHint[];
  /** Always false — slot is intent only. */
  live_dispatched: false;
}

export interface ResearchInterrogationSubagentChaseInput {
  session_id: string;
  parent_asset_id: string;
  questions: ChaseQuestion[];
  chase_mode: ChaseMode;
  would_exceed: boolean | null;
  operator_override?: boolean;
  selected_model_id?: string | null;
  source_families?: SourceFamilyHint[] | null;
  operator_ack: boolean;
  /**
   * When true, marks chase outputs as candidates for twin/session records.
   * Still does not persist (record_persisted stays false).
   */
  mark_for_twin_record?: boolean;
}

export interface ResearchInterrogationSubagentChaseCompose {
  session_id: string;
  parent_asset_id: string;
  chase_mode: ChaseMode;
  planned_slots: PlannedChaseSlot[];
  slot_count: number;
  budget_ready: boolean;
  would_exceed: boolean | null;
  mark_for_twin_record: boolean;
  /**
   * True when ≥1 planned slot, operator_ack, and budget_ready.
   * Never authorizes live dispatch.
   */
  chase_ready: boolean;
  live_dispatched: false;
  pack_dispatched: false;
  record_persisted: false;
  prompts_injected: false;
  notes: string[];
  authority: "research_interrogation_subagent_chase_compose_advisory";
}

const VALID_MODES = new Set<ChaseMode>([
  "single_question",
  "swarm_fanout",
  "collective_merge_after",
]);

const VALID_FAMILIES = new Set<SourceFamilyHint>([
  "arxiv",
  "substack",
  "openalex",
  "web",
  "custom",
]);

const SECRETISH = /sk-|api[_-]?key|secret/i;

function requireNonEmpty(value: unknown, name: string): string {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${name} must be a non-empty string`);
  }
  return value.trim();
}

/**
 * Compose a pure subagent chase plan from interrogation questions.
 * Never dispatches; never persists twin/session records; never injects prompts.
 */
export function composeResearchInterrogationSubagentChase(
  input: ResearchInterrogationSubagentChaseInput,
): ResearchInterrogationSubagentChaseCompose {
  if (!input || typeof input !== "object") {
    throw new Error("input must be an object");
  }
  if (typeof input.operator_ack !== "boolean") {
    throw new Error("operator_ack must be an explicit boolean");
  }
  if (input.would_exceed !== null && typeof input.would_exceed !== "boolean") {
    throw new Error("would_exceed must be boolean or null");
  }
  const override =
    input.operator_override === undefined ? false : input.operator_override;
  if (typeof override !== "boolean") {
    throw new Error("operator_override must be boolean when set");
  }

  const session_id = requireNonEmpty(input.session_id, "session_id");
  const parent_asset_id = requireNonEmpty(
    input.parent_asset_id,
    "parent_asset_id",
  );
  if (!VALID_MODES.has(input.chase_mode)) {
    throw new Error(
      "chase_mode must be single_question|swarm_fanout|collective_merge_after",
    );
  }
  const chase_mode = input.chase_mode;

  if (!Array.isArray(input.questions) || input.questions.length === 0) {
    throw new Error("questions must be a non-empty array");
  }

  const mark_for_twin_record =
    input.mark_for_twin_record === undefined
      ? false
      : input.mark_for_twin_record;
  if (typeof mark_for_twin_record !== "boolean") {
    throw new Error("mark_for_twin_record must be boolean when set");
  }

  const notes: string[] = [
    "live_dispatched=false — chase plan is pure intent only",
    "pack_dispatched=false — collective merge after is intent only",
    "record_persisted=false — twin/session records not written",
    "prompts_injected=false — no live prompt mutation",
  ];

  let model_id: string | null = null;
  if (input.selected_model_id != null) {
    model_id = requireNonEmpty(
      input.selected_model_id,
      "selected_model_id",
    );
    if (model_id.length > 128 || SECRETISH.test(model_id)) {
      throw new Error(
        "selected_model_id must be a model id, not secret material",
      );
    }
    notes.push(`selected_model_id=${model_id} (operator authority)`);
  } else {
    notes.push(
      "selected_model_id=null — operator may choose before live chase",
    );
  }

  const families: SourceFamilyHint[] = [];
  if (input.source_families != null) {
    if (!Array.isArray(input.source_families)) {
      throw new Error("source_families must be an array when set");
    }
    const seen = new Set<string>();
    for (let i = 0; i < input.source_families.length; i++) {
      const f = input.source_families[i];
      if (!VALID_FAMILIES.has(f)) {
        throw new Error(
          `source_families[${i}] must be arxiv|substack|openalex|web|custom`,
        );
      }
      if (seen.has(f)) {
        throw new Error(`duplicate source_family: ${f}`);
      }
      seen.add(f);
      families.push(f);
    }
  }
  notes.push(`source_family_count=${families.length}`);

  // Validate + normalize questions
  const normalized: { question_id: string; body: string; priority: number }[] =
    [];
  const seenQ = new Set<string>();
  for (let i = 0; i < input.questions.length; i++) {
    const q = input.questions[i];
    if (!q || typeof q !== "object") {
      throw new Error(`questions[${i}] must be an object`);
    }
    const question_id = requireNonEmpty(
      q.question_id,
      `questions[${i}].question_id`,
    );
    const body = requireNonEmpty(q.body, `questions[${i}].body`);
    if (seenQ.has(question_id)) {
      throw new Error(`duplicate question_id: ${question_id}`);
    }
    seenQ.add(question_id);
    let priority = 0;
    if (q.priority !== undefined) {
      if (
        typeof q.priority !== "number" ||
        !Number.isFinite(q.priority) ||
        !Number.isInteger(q.priority)
      ) {
        throw new Error(`questions[${i}].priority must be a finite integer`);
      }
      priority = q.priority;
    }
    normalized.push({ question_id, body, priority });
  }

  // Mode gates
  if (chase_mode === "single_question" && normalized.length !== 1) {
    throw new Error("single_question mode requires exactly 1 question");
  }
  if (
    (chase_mode === "swarm_fanout" ||
      chase_mode === "collective_merge_after") &&
    normalized.length < 2
  ) {
    throw new Error(`${chase_mode} requires ≥2 questions`);
  }

  // Sort by priority desc, stable by question_id
  normalized.sort((a, b) => {
    if (b.priority !== a.priority) return b.priority - a.priority;
    return a.question_id.localeCompare(b.question_id);
  });

  const planned_slots: PlannedChaseSlot[] = normalized.map((q, idx) => ({
    slot_id: `chase_${session_id}_${idx + 1}_${q.question_id}`,
    question_id: q.question_id,
    body: q.body,
    priority: q.priority,
    selected_model_id: model_id,
    source_families: [...families],
    live_dispatched: false as const,
  }));

  notes.push(
    `planned_slots=${planned_slots.length} · chase_mode=${chase_mode}`,
  );
  if (chase_mode === "collective_merge_after") {
    notes.push(
      "collective_merge_after — merge intent only; pack_dispatched=false",
    );
  }
  if (mark_for_twin_record) {
    notes.push(
      "mark_for_twin_record=true — candidates only; record_persisted=false",
    );
  }

  // Budget gate (same honesty as launch packages)
  let budget_ready = false;
  if (input.would_exceed === null) {
    if (override) {
      budget_ready = true;
      notes.push(
        "budget_ready=true via operator_override (would_exceed unknown)",
      );
    } else {
      notes.push(
        "budget_ready=false — would_exceed unknown and no operator_override",
      );
    }
  } else if (input.would_exceed === true) {
    if (override) {
      budget_ready = true;
      notes.push(
        "budget_ready=true via operator_override despite would_exceed=true",
      );
    } else {
      notes.push("budget_ready=false — would_exceed=true");
    }
  } else {
    budget_ready = true;
    notes.push("budget_ready=true — would_exceed=false");
  }

  const chase_ready =
    input.operator_ack && budget_ready && planned_slots.length > 0;
  if (!input.operator_ack) {
    notes.push("chase_ready=false — operator_ack required");
  } else if (!budget_ready) {
    notes.push("chase_ready=false — budget gate closed");
  } else {
    notes.push(
      "chase_ready=true — pure chase plan ready; still live_dispatched=false",
    );
  }

  notes.push("live_dispatched=false");
  notes.push("pack_dispatched=false");
  notes.push("record_persisted=false");
  notes.push("prompts_injected=false");

  return {
    session_id,
    parent_asset_id,
    chase_mode,
    planned_slots,
    slot_count: planned_slots.length,
    budget_ready,
    would_exceed: input.would_exceed,
    mark_for_twin_record,
    chase_ready,
    live_dispatched: false,
    pack_dispatched: false,
    record_persisted: false,
    prompts_injected: false,
    notes,
    authority: "research_interrogation_subagent_chase_compose_advisory",
  };
}

export function formatResearchInterrogationSubagentChaseSummary(
  c: ResearchInterrogationSubagentChaseCompose,
): string {
  return (
    `chase_ready=${c.chase_ready} · mode=${c.chase_mode} · ` +
    `slots=${c.slot_count} · budget_ready=${c.budget_ready} · ` +
    `live_dispatched=false · pack_dispatched=false · ` +
    `record_persisted=false · prompts_injected=false`
  );
}
